plugins/UM3NetworkPrinting/src/Cloud/MeshUploader.py

```diff
@@ -0,0 +1,148 @@
+            self._sent_bytes = 0
+            self._retries = 0
+            self._finished = False
+        self._uploadChunk()
+
+    ## Stops uploading the mesh, marking it as finished.
+    def stop(self):
+        Logger.log("i", "Stopped uploading")
+        self._finished = True
+
+    ## Uploads a chunk of the mesh to the cloud.
+    def _uploadChunk(self) -> None:
+        if self._finished:
+            raise ValueError("The upload is already finished")
+
+        first_byte, last_byte = self._chunkRange()
+        request = self._createRequest()
+
+        # now send the reply and subscribe to the results
+        self._reply = self._manager.put(request, self._data[first_byte:last_byte])
+        self._reply.finished.connect(self._finishedCallback)
+        self._reply.uploadProgress.connect(self._progressCallback)
+        self._reply.error.connect(self._errorCallback)
+
+    ## Handles an update to the upload progress
+    #  \param bytes_sent: The amount of bytes sent in the current request.
+    #  \param bytes_total: The amount of bytes to send in the current request.
+    def _progressCallback(self, bytes_sent: int, bytes_total: int) -> None:
+        Logger.log("i", "Progress callback %s / %s", bytes_sent, bytes_total)
+        if bytes_total:
+            total_sent = self._sent_bytes + bytes_sent
+            self._on_progress(int(total_sent / len(self._data) * 100))
+
+    ## Handles an error uploading.
+    def _errorCallback(self) -> None:
+        reply = cast(QNetworkReply, self._reply)
+        body = bytes(reply.readAll()).decode()
+        Logger.log("e", "Received error while uploading: %s", body)
+        self.stop()
+        self._on_error()
+
+    ## Checks whether a chunk of data was uploaded successfully, starting the next chunk if needed.
+    def _finishedCallback(self) -> None:
+        reply = cast(QNetworkReply, self._reply)
+        Logger.log("i", "Finished callback %s %s",
+                   reply.attribute(QNetworkRequest.HttpStatusCodeAttribute), reply.url().toString())
+
+        status_code = reply.attribute(QNetworkRequest.HttpStatusCodeAttribute)  # type: int
+
+        # check if we should retry the last chunk
+        if self._retries < self.MAX_RETRIES and status_code in self.RETRY_HTTP_CODES:
+            self._retries += 1
+            Logger.log("i", "Retrying %s/%s request %s", self._retries, self.MAX_RETRIES, reply.url().toString())
+            self._uploadChunk()
+            return
+
+        # Http codes that are not to be retried are assumed to be errors.
+        if status_code > 308:
+            self._errorCallback()
+            return
+
+        Logger.log("d", "status_code: %s, Headers: %s, body: %s", status_code,
+                   [bytes(header).decode() for header in reply.rawHeaderList()], bytes(reply.readAll()).decode())
+        self._chunkUploaded()
+
+    ## Handles a chunk of data being uploaded, starting the next chunk if needed.
+    def _chunkUploaded(self) -> None:
```

Comment:
Large being? Because these issues didn't occur with the cluster. This implementation can still cause issues with the GIL. 
Did you also check if the network manager from QT doesn't already support this? 
Also, if this is the reason; please document this. Because if I would have seen this at a later stage, I would have removed this in a re-factor.